File: Sources/InstantTranslator/Models/SubscriptionStatus.swift
import Foundation

enum SubscriptionStatusValue: String, Codable {
    case trial
    case active
    case pastDue = "past_due"
    case canceled
    case expired
    case unknown

    init(raw: String?) {
        switch raw?.lowercased() {
        case "active": self = .active
        case "trial", "trialing": self = .trial
        case "past_due": self = .pastDue
        case "canceled", "cancelled": self = .canceled
        case "expired": self = .expired
        default: self = .unknown
        }
    }

    /// Whether this status grants access to translation.
    var canTranslate: Bool {
        switch self {
        case .active, .trial:
            return true
        case .pastDue, .canceled, .expired, .unknown:
            return false
        }
    }
}

struct SubscriptionStatus: Codable, Equatable {
    let status: SubscriptionStatusValue
    let planName: String?
    let trialDaysRemaining: Int?

    init(status: SubscriptionStatusValue, planName: String?, trialDaysRemaining: Int?) {
        self.status = status
        self.planName = planName
        self.trialDaysRemaining = trialDaysRemaining
    }
}

/// Persisted account record stored encrypted in AccountStore.
struct StoredAccountRecord: Codable, Equatable {
    var email: String
    var sessionToken: String
    var subscriptionStatus: String?
    var planName: String?
    var lastValidatedAt: Date?
}

File: windows/ControlV.Core/Models/SubscriptionStatus.cs
using System.Text.Json.Serialization;

namespace ControlV.Core.Models;

public enum SubscriptionStatusValue { Trial, Active, PastDue, Canceled, Expired, Unknown }

public static class SubscriptionStatusValues
{
    /// Mirrors Swift `SubscriptionStatusValue(raw:)`.
    public static SubscriptionStatusValue Parse(string? raw) => raw?.ToLowerInvariant() switch
    {
        "active" => SubscriptionStatusValue.Active,
        "trial" or "trialing" => SubscriptionStatusValue.Trial,
        "past_due" => SubscriptionStatusValue.PastDue,
        "canceled" or "cancelled" => SubscriptionStatusValue.Canceled,
        "expired" => SubscriptionStatusValue.Expired,
        _ => SubscriptionStatusValue.Unknown,
    };

    public static string RawValue(this SubscriptionStatusValue value) => value switch
    {
        SubscriptionStatusValue.PastDue => "past_due",
        _ => value.ToString().ToLowerInvariant(),
    };

    public static bool CanTranslate(this SubscriptionStatusValue value) =>
        value is SubscriptionStatusValue.Active or SubscriptionStatusValue.Trial;
}

public sealed record SubscriptionStatus(SubscriptionStatusValue Status, string? PlanName, int? TrialDaysRemaining);

/// Persisted account record (encrypted at rest by the AccountStore implementation).
public sealed class StoredAccountRecord
{
    [JsonPropertyName("email")] public string Email { get; set; } = "";
    [JsonPropertyName("sessionToken")] public string SessionToken { get; set; } = "";
    [JsonPropertyName("subscriptionStatus")] public string? SubscriptionStatus { get; set; }
    [JsonPropertyName("planName")] public string? PlanName { get; set; }
    [JsonPropertyName("lastValidatedAt")] public DateTimeOffset? LastValidatedAt { get; set; }
}
